### ISSUE TITLE
Cheating possible using "show errors" and "undo"

### ISSUE THREAD

**Issue by Florikoeln** [meikpiep/gauguin#144]
It is possbible to check if a number is correct by using "show errors" (questionmark), and if "one error" is given use undo and try another number. A game where this tactic is used is not counted "used help"

Solution: either count a game as "used help" when "show errors" is used or when within a game both, "show errors" as well as "undo" are used.

**Comment by mccartney** [comment]
This tactic can also be used with ? and undo.

**Comment by meikpiep** [comment]
@Florikoeln @mccartney I think you both mean the same thing.

**Comment by meikpiep** [comment]
The point is valid and cheating is currently possible, you are right.

I personally like to play with only a few uses oft the '?' vor hinter button: If I'm unsure, I press the button. Mostly, there is no error and I continue the game. But if there are errors, I undo one or multiple moves, double check that no errors remain by using the hint button again.

It's clear that I solved such games with help, but the current ui highly suggests to use the hint button. If we change the semantics, the ui must reflect this clearly.

Maybe, a 'strict mode' without a hint button would help, but I will think about solutions.

Feel free to contribute!

**Comment by mccartney** [comment]
I think I play the same way as @meikpiep described. I use the `?` fairly often just to make sure I am not losing my time. One can say it's cheating and actually I am artificially making the game easier than it is.

Looking at it from a broader perspective I think there might be different preferences by different people, including (but likely not limited to):
1. Everything is "allowed", hints, "show errors", reveal this cage, etc. I want to have fun and not stress.
2. Let me solve the puzzle myself, but just let me check if I am on the right path.
3. Just a piece of paper and pencil. Let me figure it out myself.

My hunch is that all three options have its fans.

And since we are not having any public score board or any exchange of results, I am not sure if it matters much how we judge and what do we allow.

Maybe it matters for people who would like to switch from (2) to (3). And they would want to have the strict mode to enforce that goal.

To distinguish between (1) and (2) - I fail to have an example where it would be needed.

**Comment by meikpiep** [comment]
Thanks @mccartney, makes perfect sense to me, great to read how you sort this out.

I will think about how to support these player types effective. I'm not sure if it is just a 'no hint button' option or something else.

**Comment by meikpiep** [comment]
@Florikoeln What do you say - would a kind of mode suffice where the hint button is totally invisible or reacts after a warning only and marks a game as cheated as you described it? Any ideas?

**Comment by mccartney** [comment]
Thinking about it more. Maybe it's just enough to add a setting (off by default): "consider ? as cheating" (or whatever nice term we want to use).

**Comment by Florikoeln** [comment]
> @Florikoeln What do you say - would a kind of mode suffice where the hint button is totally invisible or reacts after a warning only and marks a game as cheated as you described it? Any ideas?

The only impact of using "?" and "undo" could be the statistics (the game is counted as solved without help). 

I as a strict referee I would say that using "?" is help. But I agree, that it is very useful (and often used by me). Therefore I would agree to @mccartney to add an option "consider ? help" (plus maybe "consider undo as help") which could be off by default.

**Comment by KranzAklilu** [comment]
@Florikoeln I agree with your suggestion and I think adding a row on the statistic page for (Puzzles solved without help) would be a great addition.
Another suggestion would be to limit the amount of times you could use the ? sign per puzzle. as it'd encourage not to use it every time. I'd argue it would be more fun and rewarding when solving puzzles.

**Comment by DancingGeek** [comment]
I'd like to share my own perspective on the undo functionality, because I don't see it as directly helping to solve the puzzle, but rather as a quality of life feature.

If I don't let the game tell me when something is wrong, then undo is dependent on my awareness of having made a mistake and when I made it. However, it also lets me catch moments where I mistype something, perhaps I realise I added a wrong number as a possible number but I'm not sure which one and don't want to have to start from scratch in that square (not always necessary, but sometimes it would be a real pain). Maybe I accidentally long pressed rather than short pressed, and don't remember what was in the square before. Undo let's me browse back in time through my changes, but doesn't tell me if I made a mistake.

The hint button is definitely helping me with solving the puzzle, so if using that meant it cleared my streak it would feel fair. But if that's the case, I would want it far away from the number pad, perhaps in the same menu as Reveal cell for example. Even with the safety of the 'Are you sure?' dialogue box, I don't want to accidentally click it. A pop-up warning is disruptive if I'm accidentally clicking it, and way too easy to accidentally click yes to if I'm tapping away furiously trying to beat my best time score.

Or maybe I'm too attached to my streak?
